Automattic/wp-calypso
# Add async toast notification component

Calypso uses toast notifications to present information with a limited lifespan, like fixable errors in form input or alerts about failed operations from the wpcom REST API.

<img width="350" alt="Screenshot 2023-10-24 at 2 32 33 PM" src="https://github.com/Automattic/wp-calypso/assets/9310939/aa77667f-ce93-4f0f-b30b-303ff05f0787">

These are preferable to bell notifications for information that is immediate and ephemeral. However unlike bell notifications we have no facility on the backend for triggering toast notifications asynchronously; if an API call is such that we want to show a toast, it includes this in the response to the client and we handle it ad-hoc.

There are situations where work on the backend happens in an async job but we'd like to alert the user to the outcome. D125318-code adds a backend API and two REST endpoints for creating and fetching async toasts; this patch adds the state and components needed to use these in calypso.

Reviewers: I tried to put the commits in an order that makes sense.

## Proposed Changes

* The backend API presents an endpoint that returns a mapping from string keys to a message and severity; this mapping is associated to a specific site.
* This PR adds a higher order component, `withAsyncToast`. It takes an array of keys as a parameter and attaches a listener to the wrapped component that polls the API and displays the keyed message exactly once.

Polling is expensive, but the backend API is only accessible to site admins and these messages are meant only to be used in admin screens.

## Testing Instructions

Unit tests: `yarn run test-client async-toast`

Manual test:
* To test this, you will need to sandbox `public-api.wordpress.com` with D125318-code applied.
* Log in as a test user and open wp-admin for a site which has a purchase. Let $siteId be the ID of this site (you'll need this later).
* Open devtools and filter for `async-toast` in the network tab.
* Navigate to the site purchases page in Upgrades > Purchases; the path is `/purchases/subscriptions/$siteUrl`.
* Click to select an individual purchase. Observe a GET request to `https://public-api.wordpress.com/wpcom/v2/sites/$siteId/async-toast` with response like this:

<img width="408" alt="Screenshot 2023-10-26 at 10 23 42 PM" src="https://github.com/Automattic/wp-calypso/assets/9310939/90d49455-194b-4664-b722-38d137ecaf89">

* On your wpcom sandbox, open `wpsh`. Invoke this command:

```
Async_Toast::add_new_toast( $siteId, 'test-toast', 'error', 'all your base are belong to us' );
```

* In the browser, navigate back to the site purchases screen, and then select an individual purchase. Observe a toast notification:

<img width="388" alt="Screenshot 2023-10-26 at 10 26 37 PM" src="https://github.com/Automattic/wp-calypso/assets/9310939/dddf3735-76da-4c7d-b000-31b53e79d05f">

* In the network tab, notice a call to `/sites/$siteId/async-toast` with a nonempty response, followed by a DELETE call to `/sites/$siteId/async-toast/test-toast`, followed by another call to `/sites/$siteId/async-toast` now with an empty response.

* The notification should appear only once. Try clicking to the site purchases list and back again; check that the notice does not fire a second time. Refresh the purchase management screen; the notice should not fire.

## Pre-merge Checklist
